# Final_Project_BIO539

This project includes a pipeline for RNA seq data analysis from the SRA file raw reads stage to differential gene expression and principal component analysis in R. 
The main software used in this pipeline are linux codes on the HPC, SRA toolkit, Fastqc, STAR, Stringtie, bash, python, R ->Deseq2, ggplot, tidyverse. 

I have also uploaded to this repository my work/notes on this project as an R markdwon file (and it's pdf), as well as the final project essay as a word document.

Here are the references I used when doing this project:
1. Asthma Facts and Figures. Asthma and Allergy Foundation of America. Available from: https://www.aafa.org/asthma-facts/#:~:text=According%20to%20the%20Centers%20for,in%2013%20people%20have%20asthma.&text=More%20than%2025%20million%20Americans,age%2C%20sex%20and%20racial%20groups.
2. Count-Based Differential Expression Analysis of RNA-seq Data. Available from: https://4va.github.io/biodatasci/r-rnaseq-airway.html#pca
3. Dobin, A., & Gingeras, T. R. (2015). Mapping RNA-seq Reads with STAR. Current protocols in bioinformatics, 51, 11.14.1â€“11.14.19. https://doi.org/10.1002/0471250953.bi1114s51
4. Fastqc. Available from : https://www.bioinformatics.babraham.ac.uk/projects/fastqc/
5. GEO NCBI. Human Airway Smooth Muscle Transcriptome Changes in Response to Asthma Medications. Available from : https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE52778
6. GRCh38. Available from: ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_37/GRCh38.p13.genome.fa.gz
7. GRCh38 Annotation. Available from: ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_37/gencode.v37.chr_patch_hapl_scaff.basic.annotation.gtf.gz 
8. Himes BE, Jiang X, Wagner P, Hu R et al. RNA-Seq transcriptome profiling identifies CRISPLD2 as a glucocorticoid responsive gene that modulates cytokine function in airway smooth muscle cells. PLoS One 2014;9(6):e99625. PMID: 24926665
9. Roberts, E.M. 2020. "Pipeline to perform differential expression analysis with transcriptomes". https://github.com/erinroberts/apoptosis_data_pipeline. The University of Rhode Island.
10. SRA Toolkit. Available from : https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/2.11.0/sratoolkit.2.11.0-centos_linux64.tar.gz
11. STAR. Available from : https://github.com/alexdobin/STAR
12. Stringtie. Available from : https://ccb.jhu.edu/software/stringtie/index.shtml?t=manual	
13. Stringtie Python prepDE Script : Available from : https://ccb.jhu.edu/software/stringtie/dl/prepDE.py
